cluster-autoscaler/core/filteroutschedulable/filter_out_schedulable.go

```diff
@@ -97,94 +94,26 @@ func (p *filterOutSchedulablePodListProcessor) CleanUp() {
-	unschedulablePodsCache := utils.NewPodSchedulableMap()
-
+func (p *filterOutSchedulablePodListProcessor) filterOutSchedulableByPacking(unschedulableCandidates []*apiv1.Pod, clusterSnapshot clustersnapshot.ClusterSnapshot) ([]*apiv1.Pod, error) {
 	// Sort unschedulable pods by importance
 	sort.Slice(unschedulableCandidates, func(i, j int) bool {
-		return moreImportantPod(unschedulableCandidates[i], unschedulableCandidates[j])
+		return corev1helpers.PodPriority(unschedulableCandidates[i]) > corev1helpers.PodPriority(unschedulableCandidates[j])
 	})
 
 	// Pods which remain unschedulable
 	var unschedulablePods []*apiv1.Pod
-
-	// Try to schedule based on hints
-	podsFilteredUsingHints := 0
-	podsToCheckAgainstAllNodes := make([]*apiv1.Pod, 0, len(unschedulableCandidates))
-	for _, pod := range unschedulableCandidates {
-		scheduledOnHintedNode := false
-		if hintedNodeName, hintFound := p.schedulablePodsNodeHints[pod.UID]; hintFound {
-			if predicateChecker.CheckPredicates(clusterSnapshot, pod, hintedNodeName) == nil {
-				// We treat predicate error and missing node error here in the same way
-				scheduledOnHintedNode = true
-				podsFilteredUsingHints++
-				klog.V(4).Infof("Pod %s.%s marked as unschedulable can be scheduled on node %s (based on hinting). Ignoring"+
-					" in scale up.", pod.Namespace, pod.Name, hintedNodeName)
-
-				if err := clusterSnapshot.AddPod(pod, hintedNodeName); err != nil {
-					return nil, err
-				}
-			}
-		}
-
-		if !scheduledOnHintedNode {
-			podsToCheckAgainstAllNodes = append(podsToCheckAgainstAllNodes, pod)
-			delete(p.schedulablePodsNodeHints, pod.UID)
-		}
-	}
-	klog.V(4).Infof("Filtered out %d pods using hints", podsFilteredUsingHints)
-
-	// Cleanup hints map
-	foundPods := make(map[types.UID]bool)
 	for _, pod := range unschedulableCandidates {
-		foundPods[pod.UID] = true
-	}
-	for hintedPodUID := range p.schedulablePodsNodeHints {
-		if !foundPods[hintedPodUID] {
-			delete(p.schedulablePodsNodeHints, hintedPodUID)
-		}
-	}
-
-	// Try to bin pack remaining pods
-	unschedulePodsCacheHitCounter := 0
-	for _, pod := range podsToCheckAgainstAllNodes {
-		_, found := unschedulablePodsCache.Get(pod)
-		if found {
-			// Cache hit for similar pod; assuming unschedulable without running predicates
-			unschedulablePods = append(unschedulablePods, pod)
-			unschedulePodsCacheHitCounter++
-			continue
-		}
-		nodeName, err := predicateChecker.FitsAnyNode(clusterSnapshot, pod)
-		if err == nil {
-			klog.V(4).Infof("Pod %s.%s marked as unschedulable can be scheduled on node %s. Ignoring"+
-				" in scale up.", pod.Namespace, pod.Name, nodeName)
-			if err := clusterSnapshot.AddPod(pod, nodeName); err != nil {
-				return nil, err
-			}
-			// Store hint for pod placement
-			p.schedulablePodsNodeHints[pod.UID] = nodeName
-		} else {
+		_, err := p.schedulingSimulator.TrySchedulePods(clusterSnapshot, []*apiv1.Pod{pod}, scheduling.ScheduleAnywhere)
```

Comment:
Maybe instead of looping over pods here it would be possible to return some `SchedulingStatus` struct as a first argument and have unschedulable pods in it as a field?

Reply:
I've added a `scheduling.Status` struct that contains:
- the scheduled pod
- the node name

I've also needed to change the `TrySchedulePods` logic - up until now it failed if there was an unexpected error OR if some pod was unschedulable. Now it fails only on unexpected errors and returns the subset of scheduled pods if it does not manage to schedule all.

This required some changes around calls to `TrySchedulePods`, but reproducing the previous behaviour can be done just by comparing the length of the pod list vs the length of the scheduled pods list.

Reply:
Comparing list lengths makes sense from the correctness point of view, but I'd consider having a bool param in scheduling simulator (ctor) whether to error out early on first scheduling error - this may have significant impact on scale down performance.

Reply:
(Alternatively: a param to `TrySchedulePods`)

Reply:
Added `scheduleAll` param to the `TrySchedulePods` - this should help from the performance perspective. 

I don't think however it would be good to return an error if we want to schedule all. Having a function which return two different formats depending on the boolean flag is confusing. So I left the length comparison I previously added, and just added a `break` to the `TrySchedulePods`.